account_mtd/models/mtd_hello_world.py

```diff
@@ -0,0 +1,456 @@
+                path=self.path
+            )
+        )
+        return version
+
+    def _handle_mtd_hello_user_endpoint(self):
+        self.which_button_type_clicked = "user"
+        self.path = "/hello/user"
+        self._logger.info(
+            "Connection button Clicked - endpoint name {name}, redirect URL:- {redirect}, Path url:- {path}".format(
+                name=self.name,
+                redirect=self.hmrc_configuration.redirect_url,
+                path=self.path
+            )
+        )
+        # search for token record for the API
+        token_record = self.env['mtd.api_tokens'].search([('api_id', '=', self.api_name.id)])
+        self._logger.info(
+            "Connection button Clicked - endpoint name {name}, and the api is :- {api_name} ".format(
+                name=self.name,
+                api_name=self.api_name
+            )
+        )
+
+        if token_record.access_token and token_record.refresh_token:
+            self._logger.info(
+                "Connection button Clicked - endpoint name {name}, ".format(name=self.name) +
+                "We have access token and refresh token"
+            )
+            version = self._json_command('version')
+            return version
+        else:
+            self._logger.info(
+                "Connection button Clicked - endpoint name {name}, No access token ".format(name=self.name) +
+                "found and no refresh_token found from the token record table."
+            )
+            authorisation_tracker = self.env['mtd.api_request_tracker'].search([('closed', '=', False)])
+            self._logger.info(
+                "Connection button Clicked - endpoint name {name}, ".format(name=self.name) +
+                "Checking to see if a request is in process"
+            )
+            create_date = fields.Datetime.from_string(authorisation_tracker.create_date)
+            time_10_mins_ago = (datetime.utcnow() - timedelta(minutes=10))
+            if authorisation_tracker:
+                authorised_code_expired = create_date <= time_10_mins_ago
+                if authorised_code_expired:
+                    # we can place a new request and close this request.
+                    authorisation_tracker.closed = 'timed_out'
+                    self._logger.info(
+                        "Connection button Clicked - endpoint name {name}, no Pending requests".format(name=self.name)
+                    )
+                    return self.get_user_authorisation()
+                else:
+                    # The request made was within 10 mins so the user has to wait.
+                    raise exceptions.Warning(
+                        "An authorisation request is already in process!!!\n " +
+                        "Please try again later"
+                    )
+            else:
+                return self.get_user_authorisation()
+
+    def _json_command(self, command, timeout=3, record_id=None):
```

Comment:
These are pretty hefty functions. Is  there any way that these could be reduced, like having common shared functionality (it may not be shared yet, but might be in future?) split out into smaller functions?

Reply:
Doesn't need to be shared functionality to be worthy of being in a method, but common stuff (as long as it's not only common by coincidence) should be.  If something is a sequence of things that need to be done.

```
_this()
_that()
_the_other()
```

and each of those functions explain *how* to do this, that and the other, it's much easier to read.  Just be careful not to mix levels of abstraction.